pkg/controller/util/handlers.go

```diff
@@ -31,6 +31,10 @@ import (
 func NewTriggerOnAllChanges(triggerFunc func(pkgruntime.Object)) *cache.ResourceEventHandlerFuncs {
 	return &cache.ResourceEventHandlerFuncs{
 		DeleteFunc: func(old interface{}) {
+			if deleted, ok := old.(cache.DeletedFinalStateUnknown); ok {
+				// This object might be stale but ok for our current usage
+				old = deleted.Obj
+			}
 			oldObj := old.(pkgruntime.Object)
```

Comment:
Ugh, this code is terrible.  Would you be up for ensuring that all handlers only call `triggerFunc` for a non-nil (successfully cast) `pkgruntime.Object` in a subsequent PR?  All of the handlers passed as triggerFunc are assuming this.

Reply:
can you elaborate what you mean?
i know the assumption and it's why i made this change.

Reply:
The deletion handler should check that the object provided is non-nil before calling triggerFunc:

```golang
if oldObj != nil {
   triggerFunc(oldObj)
}
```

Reply:
i don't think it can be nil.

Reply:
Please review the delta fifo code responsible for generating events of type `DeletedFinalStateUnknown`:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L474

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L502

In both cases, the deleted object may not be present and may be passed as nil in the event.

Reply:
in both cases, i don't see how it happens without bugs.
but i guess it's better to be safe than sorry. i'll add a nil check.